src/Tempest/Core/src/Kernel.php

```diff
@@ -76,29 +81,48 @@ private function createContainer(): Container
         return $container;
     }
 
-    private function loadComposer(): self
+    public function loadComposer(): self
     {
         $this->container->singleton(Composer::class, new Composer($this->root));
 
         return $this;
     }
 
-    private function loadEnv(): self
+    public function loadEnv(): self
     {
+        // The discovery:generate command must always run without discovery cache enabled
+        if (PHP_SAPI === 'cli') {
+            $command = $_SERVER['argv'][1] ?? null;
+
+            if ($command === 'discovery:generate' || $command === 'dg') {
+                putenv('DISCOVERY_CACHE=false');
+            }
+        }
+
+        // Load from .env
         $dotenv = Dotenv::createUnsafeImmutable($this->root);
         $dotenv->safeLoad();
 
+        // Check whether the discovery caching strategy in .env has changed
+        $current = env('DISCOVERY_CACHE');
+
+        if ($current && $original = @file_get_contents(DiscoveryGenerateCommand::CURRENT_DISCOVERY_STRATEGY)) {
```

Comment:
I don't like storing the currently generated discovery cache strategy in a file. But I have no idea how else I could handle it. 

If we don't have this check, Tempest will fail _somewhere_ whenever you change your environment variable (most likely when you first try to resolve a dependency from the container).

So I rather have a clear exception being thrown telling the user they should rerun `discovery:generate`.

On the other hand… maybe it isn't a big deal? I doubt anyone will often change their discovery cache config…